site_forecast_app/data/satellite.py

```diff
@@ -0,0 +1,143 @@
+"""Download and unzip satellite data."""
+import logging
+import os
+import tempfile
+import zipfile
+from datetime import UTC, datetime, timedelta
+
+import fsspec
+import numpy as np
+import xarray as xr
+
+log = logging.getLogger(__name__)
+
+def satellite_scale_minmax(ds: xr.Dataset) -> xr.Dataset:
+    """Scale the satellite dataset via min-max to [0,1] range."""
+    log.info("Scaling satellite data to 0,1] range via min-max")
+
+    channels = ds.variable.values
+    # min and max values for each variable (same length as `variable`
+    # and in the same order)
+    min_vals = np.array(
+            [
+                -2.5118103,
+                -64.83977,
+                63.404694,
+                2.844452,
+                199.10002,
+                -17.254883,
+                -26.29155,
+                -1.1009827,
+                -2.4184198,
+                199.57048,
+                198.95093,
+            ])
+    max_vals = np.array(
+            [
+                69.60857,
+                339.15588,
+                340.26526,
+                317.86752,
+                313.2767,
+                315.99194,
+                274.82297,
+                93.786545,
+                101.34922,
+                249.91806,
+                286.96323,
+            ])
+
+    # Create DataArrays for min and max with the 'variable' dimension
+    min_da = xr.DataArray(min_vals, coords={"variable": channels}, dims=["variable"])
+    max_da = xr.DataArray(max_vals, coords={"variable": channels}, dims=["variable"])
+
+    # Apply scaling
+    scaled_ds = (ds - min_da) / (max_da - min_da)
+    scaled_ds = scaled_ds.clip(min=0, max=1)  # Ensure values are within [0, 1]
+    return scaled_ds
+
+
+def download_satellite_data(satellite_source_file_path: str,
+                            local_satellite_path: str,
+                            scaling_method: str = "constant",
+                            satellite_backup_source_file_path: None | str = None) -> None:
+    """Download the sat data."""
+    if os.path.exists(local_satellite_path):
+        log.info(f"File already exists at {local_satellite_path}")
+        return
+
+    with tempfile.TemporaryDirectory() as tmpdir:
+
+        temporary_satellite_data = f"{tmpdir}/temporary_satellite_data.zarr"
+
+        download_and_unzip(file_zip=satellite_source_file_path, file=temporary_satellite_data)
+
+        # log the timestamps for satellite data
+        ds = xr.open_zarr(temporary_satellite_data)
+        times = ds.time.values
+        log.info(f"Satellite data timestamps: {times}")
+
+        # possibily download backup satellite
+        # if there are not enough time in the current satellite data
+        latest_satellite_time = times.max()
+        now = datetime.now(UTC)
+        latest_satellite_time = datetime.fromtimestamp(
+            int(latest_satellite_time.astype(datetime)/1e9), tz=UTC)
```

Comment:
I could change this to use pandas, and then we wouldnt have this conversation here